In [143]:
import pandas as pd
import numpy as np

In [144]:
df=pd.read_csv("./final (2).csv")

In [145]:
df.drop(['Unnamed: 0'],inplace=True,axis=1)

In [146]:
df

FL_DATE ORIGIN DEST  CRS_DEP_TIME  DEP_TIME  DEP_DELAY  WHEELS_OFF  \
0       2018-01-01    ATL  EWR          1819    1818.0       -1.0      1834.0   
1       2018-01-01    ATL  SFO           730     729.0       -1.0       750.0   
2       2018-01-01    ATL  EWR          1555    1549.0       -6.0      1626.0   
3       2018-01-01    ATL  ORD          1748    1751.0        3.0      1808.0   
4       2018-01-01    ATL  IAD           943    1005.0       22.0      1023.0   
...            ...    ...  ...           ...       ...        ...         ...   
387570  2018-12-31    ATL  CLT          1220    1240.0       20.0      1300.0   
387571  2018-12-31    ATL  PHX          1125    1224.0       59.0      1236.0   
387572  2018-12-31    ATL  DFW          1419    1413.0       -6.0      1428.0   
387573  2018-12-31    ATL  ORD          1737    1740.0        3.0      1758.0   
387574  2018-12-31    ATL  CLT           659     659.0        0.0       715.0   

        WHEELS_ON  TAXI_IN  CRS_ARR_TIME  ARR_DELAY  CANCELLED  DIVERTED  \
0          2004.0     17.0          2048      -27.0        0.0       0.0   
1           948.0      4.0           956       -4.0        0.0       0.0   
2          1826.0     29.0          1816       39.0        0.0       0.0   
3          1843.0      3.0          1901      -15.0        0.0       0.0   
4          1133.0      3.0          1130        6.0        0.0       0.0   
...           ...      ...           ...        ...        ...       ...   
387570     1356.0      6.0          1337       25.0        0.0       0.0   
387571     1447.0      6.0          1345       68.0        0.0       0.0   
387572     1525.0     14.0          1549      -10.0        0.0       0.0   
387573     1830.0     31.0          1904       -3.0        0.0       0.0   
387574      751.0     19.0           818       -8.0        0.0       0.0   

        CRS_ELAPSED_TIME  ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  \
0                  149.0                123.0      90.0     746.0   
1                  326.0                323.0     298.0    2139.0   
2                  141.0                186.0     120.0     746.0   
3                  133.0                115.0      95.0     606.0   
4                  107.0                 91.0      70.0     534.0   
...                  ...                  ...       ...       ...   
387570              77.0                 82.0      56.0     226.0   
387571             260.0                269.0     251.0    1587.0   
387572             150.0                146.0     117.0     731.0   
387573             147.0                141.0      92.0     606.0   
387574              79.0                 71.0      36.0     226.0   

        CARRIER_DELAY  WEATHER_DELAY  LATE_AIRCRAFT_DELAY  Temperature  \
0                 NaN            NaN                  NaN           25   
1                 NaN            NaN                  NaN           19   
2                 0.0            0.0                  0.0           28   
3                 NaN            NaN                  NaN           27   
4                 NaN            NaN                  NaN           19   
...               ...            ...                  ...          ...   
387570            0.0            0.0                 20.0           66   
387571            4.0            0.0                 55.0           66   
387572            NaN            NaN                  NaN           68   
387573            NaN            NaN                  NaN           70   
387574            NaN            NaN                  NaN           56   

        Dew Point  Humidity  Wind Speed  Wind Gust  Pressure  Precipitation  \
0               8        49          12          0     29.36            0.0   
1               5        54          13          0     29.27            0.0   
2              11        49           9          0     29.33            0.0   
3              10        49          13          0     29.35            0.0   
4           

In [147]:
pd.set_option('display.max_columns',None)

In [148]:
df.drop(['WHEELS_OFF','WHEELS_ON','TAXI_IN','DEP_TIME','CRS_ARR_TIME','ARR_DELAY',
        'CRS_ELAPSED_TIME','ACTUAL_ELAPSED_TIME','Wind Gust'],inplace=True,axis=1)

In [149]:
df.head(20)

FL_DATE ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  CANCELLED  DIVERTED  \
0   2018-01-01    ATL  EWR          1819       -1.0        0.0       0.0   
1   2018-01-01    ATL  SFO           730       -1.0        0.0       0.0   
2   2018-01-01    ATL  EWR          1555       -6.0        0.0       0.0   
3   2018-01-01    ATL  ORD          1748        3.0        0.0       0.0   
4   2018-01-01    ATL  IAD           943       22.0        0.0       0.0   
5   2018-01-01    ATL  EWR          2011       45.0        0.0       0.0   
6   2018-01-01    ATL  EWR          1016       -1.0        0.0       0.0   
7   2018-01-01    ATL  EWR          1329        1.0        0.0       0.0   
8   2018-01-01    ATL  EWR          1122       -6.0        0.0       0.0   
9   2018-01-01    ATL  EWR          2040       -2.0        0.0       0.0   
10  2018-01-01    ATL  SFO          1931       -4.0        0.0       0.0   
11  2018-01-01    ATL  IAD          1420       -5.0        0.0       0.0   
12  2018-01-01    ATL  IAH          1738        9.0        0.0       0.0   
13  2018-01-01    ATL  SEA           617        6.0        0.0       0.0   
14  2018-01-01    ATL  SEA          1600       -4.0        0.0       0.0   
15  2018-01-01    ATL  DSM           950       68.0        0.0       0.0   
16  2018-01-01    ATL  JAN          1215       -4.0        0.0       0.0   
17  2018-01-01    ATL  ORD          1525        4.0        0.0       0.0   
18  2018-01-01    ATL  ABY          2036       -5.0        0.0       0.0   
19  2018-01-01    ATL  AEX          2205       11.0        0.0       0.0   

    AIR_TIME  DISTANCE  CARRIER_DELAY  WEATHER_DELAY  LATE_AIRCRAFT_DELAY  \
0       90.0     746.0            NaN            NaN                  NaN   
1      298.0    2139.0            NaN            NaN                  NaN   
2      120.0     746.0            0.0            0.0                  0.0   
3       95.0     606.0            NaN            NaN                  NaN   
4       70.0     534.0            NaN            NaN                  NaN   
5       94.0     746.0            NaN            NaN                  NaN   
6       88.0     746.0            NaN            NaN                  NaN   
7      132.0     746.0            0.0            0.0                  0.0   
8       86.0     746.0            NaN            NaN                  NaN   
9       90.0     746.0            NaN            NaN                  NaN   
10     291.0    2139.0            NaN            NaN                  NaN   
11      70.0     534.0            NaN            NaN                  NaN   
12     106.0     689.0            NaN            NaN                  NaN   
13     312.0    2182.0            NaN            NaN                  NaN   
14     313.0    2182.0            NaN            NaN                  NaN   
15     113.0     743.0            0.0           68.0                  0.0   
16      57.0     341.0            NaN            NaN                  NaN   
17      91.0     606.0            NaN            NaN                  NaN   
18      30.0     145.0            NaN            NaN                  NaN   
19      76.0     500.0            NaN            NaN                  NaN   

    Temperature  Dew Point  Humidity  Wind Speed  Pressure  Precipitation  \
0            25          8        49          12     29.36            0.0   
1            19          5        54          13     29.27            0.0   
2            28         11        49           9     29.33            0.0   
3            27         10        49          13     29.35            0.0   
4            19          6        57          15     29.36            0.0   
5             0          0         0           0      0.00            0.0   
6            19          6        57          15     29.36            0.0   
7            23          8        53          13     29.34            0.0   
8            20          7        57          10     29.37            0.0   
9             0       

In [150]:
df.CANCELLED.value_counts()

0.0    387502
1.0        73
Name: CANCELLED, dtype: int64

In [151]:
df.DIVERTED.value_counts()

0.0    386659
1.0       916
Name: DIVERTED, dtype: int64

In [152]:
total_flights_delayed=df[(df['DEP_DELAY']>0)]
df=df[(df['DEP_DELAY']>0)]

In [153]:
weather_delayed_flights=df[(df['WEATHER_DELAY']>0)]

# Ratio of weather delayed flights

In [154]:
print("Out of {} no. of flights delayed, {} no. of flights were delayed due to weather conditions".format(len(total_flights_delayed),
                                                                                                len(weather_delayed_flights)))

Out of 141182 no. of flights delayed, 8408 no. of flights were delayed due to weather conditions


In [155]:
(len(weather_delayed_flights)/len(total_flights_delayed))*100

5.955433412191356

In [156]:
df.Condition.value_counts()

Mostly Cloudy              43441
Fair                       24144
Cloudy                     20170
Partly Cloudy              17887
Light Rain                 10850
0                          10627
Fog                         2159
Light Drizzle               2116
Rain                        1491
Fair / Windy                1093
Thunder in the Vicinity      746
T-Storm                      708
Light Rain with Thunder      701
Mostly Cloudy / Windy        700
Drizzle and Fog              685
Cloudy / Windy               643
Heavy Rain                   642
Thunder                      532
Heavy T-Storm                471
Haze                         336
Mist                         244
Partly Cloudy / Windy        187
Light Rain / Windy           164
Rain / Windy                 160
Heavy T-Storm / Windy         98
Light Snow                    80
Heavy Rain / Windy            67
Patches of Fog                40
Name: Condition, dtype: int64

In [157]:
df.DEP_DELAY.value_counts()

1.0       10303
2.0        7602
3.0        6697
4.0        6179
5.0        5607
          ...  
561.0         1
846.0         1
681.0         1
374.0         1
1206.0        1
Name: DEP_DELAY, Length: 720, dtype: int64

In [158]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 141182 entries, 3 to 387573
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   FL_DATE              141182 non-null  object 
 1   ORIGIN               141182 non-null  object 
 2   DEST                 141182 non-null  object 
 3   CRS_DEP_TIME         141182 non-null  int64  
 4   DEP_DELAY            141182 non-null  float64
 5   CANCELLED            141182 non-null  float64
 6   DIVERTED             141182 non-null  float64
 7   AIR_TIME             140685 non-null  float64
 8   DISTANCE             141182 non-null  float64
 9   CARRIER_DELAY        58711 non-null   float64
 10  WEATHER_DELAY        58711 non-null   float64
 11  LATE_AIRCRAFT_DELAY  58711 non-null   float64
 12  Temperature          141182 non-null  int64  
 13  Dew Point            141182 non-null  int64  
 14  Humidity             141182 non-null  int64  
 15  Wind Speed       

In [159]:
df.isnull().sum()

FL_DATE                    0
ORIGIN                     0
DEST                       0
CRS_DEP_TIME               0
DEP_DELAY                  0
CANCELLED                  0
DIVERTED                   0
AIR_TIME                 497
DISTANCE                   0
CARRIER_DELAY          82471
WEATHER_DELAY          82471
LATE_AIRCRAFT_DELAY    82471
Temperature                0
Dew Point                  0
Humidity                   0
Wind Speed                 0
Pressure                   0
Precipitation              0
Condition                  0
time                       0
dtype: int64

In [160]:
df['AIR_TIME'].fillna(df['AIR_TIME'].mean,inplace=True)
df['CARRIER_DELAY'].fillna(int(0),inplace=True)
df['WEATHER_DELAY'].fillna(int(0),inplace=True)
df['LATE_AIRCRAFT_DELAY'].fillna(int(0),inplace=True)

/home/devesh/.local/lib/python3.8/site-packages/pandas/core/series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [161]:
df.isnull().sum()

FL_DATE                0
ORIGIN                 0
DEST                   0
CRS_DEP_TIME           0
DEP_DELAY              0
CANCELLED              0
DIVERTED               0
AIR_TIME               0
DISTANCE               0
CARRIER_DELAY          0
WEATHER_DELAY          0
LATE_AIRCRAFT_DELAY    0
Temperature            0
Dew Point              0
Humidity               0
Wind Speed             0
Pressure               0
Precipitation          0
Condition              0
time                   0
dtype: int64

In [162]:
# df=df[df['DEP_DELAY','CANCELLED','DIVERTED','AIR_TIME','DISTANCE'].astype(int)]
df = df.astype({"DEP_DELAY":'int64', "CANCELLED":'int64',"DIVERTED":'int64',"DISTANCE":'int64',"CARRIER_DELAY":'int64',"WEATHER_DELAY":'int64',"LATE_AIRCRAFT_DELAY":'int64'})

In [163]:
df

FL_DATE ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  CANCELLED  DIVERTED  \
3       2018-01-01    ATL  ORD          1748          3          0         0   
4       2018-01-01    ATL  IAD           943         22          0         0   
5       2018-01-01    ATL  EWR          2011         45          0         0   
7       2018-01-01    ATL  EWR          1329          1          0         0   
12      2018-01-01    ATL  IAH          1738          9          0         0   
...            ...    ...  ...           ...        ...        ...       ...   
387563  2018-12-31    ATL  ILM          2119          2          0         0   
387569  2018-12-31    ATL  PHX          1740         76          0         0   
387570  2018-12-31    ATL  CLT          1220         20          0         0   
387571  2018-12-31    ATL  PHX          1125         59          0         0   
387573  2018-12-31    ATL  ORD          1737          3          0         0   

       AIR_TIME  DISTANCE  CARRIER_DELAY  WEATHER_DELAY  LATE_AIRCRAFT_DELAY  \
3            95       606              0              0                    0   
4            70       534              0              0                    0   
5            94       746              0              0                    0   
7           132       746              0              0                    0   
12          106       689              0              0                    0   
...         ...       ...            ...            ...                  ...   
387563       59       377              0              0                    0   
387569      240      1587             75              0                    0   
387570       56       226              0              0                   20   
387571      251      1587              4              0                   55   
387573       92       606              0              0                    0   

        Temperature  Dew Point  Humidity  Wind Speed  Pressure  Precipitation  \
3                27         10        49          13     29.35            0.0   
4                19          6        57          15     29.36            0.0   
5                 0          0         0           0      0.00            0.0   
7                23          8        53          13     29.34            0.0   
12               27         10        49          13     29.35            0.0   
...             ...        ...       ...         ...       ...            ...   
387563           67         61        81          14     28.94            0.0   
387569           68         59        73          13     28.91            0.0   
387570           66         63        90           9     28.94            0.1   
387571           66         63        90           9     28.94            0.1   
387573           70         59        68          17     28.92            0.0   

            Condition                 time  
3       Partly Cloudy  2018-01-01 17:51:00  
4          Light Snow  2018-01-01 10:05:00  
5                   0  2018-01-01 20:56:00  
7              Cloudy  2018-01-01 13:30:00  
12      Partly Cloudy  2018-01-01 17:47:00  
...               ...                  ...  
387563  Mostly Cloudy  2018-12-31 21:21:00  
387569  Mostly Cloudy  2018-12-31 18:56:00  
387570     Light Rain  2018-12-31 12:40:00  
387571     Light Rain  2018-12-31 12:24:00  
387573  Mostly Cloudy  2018-12-31 17:40:00  

[141182 rows x 20 columns]

#  Understanding Passenger Carrier Delay Costs

The **"Total Airline Operating Cost Breakdown"** consists of:
<br>
<br>
- 44% aircraft operating expenses, which include fuel, direct maintenance, 
  depreciation, and crew
<br>
<br>
- 29% servicing expense:
    - Aircraft servicing (7%)
    - Traffic servicing (11%)
    - Passenger service (11%)
<br>
<br>
- 14% reservations and sales expense
    - This figure was 19.5% in 1993, but declined steadily throughout the 1990s 
<br>
<br>
- 13% overhead expense
    - Advertising and Publicity (2%)
    - General and Administrative (6%)
    - Miscellaneous expense (5%)

The **"Direct Aircraft Operating Cost per Block Minute"** for 2018:
<br>
<br>
<ul>
    <li>Fuel : 27.01</li>
    <li>Pilots/Flight Attendants : 23.35</li>
    <li>Maintenance : 11.75</li>
    <li>Aircraft Ownership : 9.28</li>
    <li>Other : 2.79</li>
</ul>
    <br>
<br>
<strong>Total Direct Operating Costs per Minute</strong> : 74.18
<br>
<br>
<strong>Note</strong>: costs based on DOT Form 41 data for U.S. scheduled passenger airlines

<strong>Additional Costs</strong>
<br>
<br>

Delayed aircraft are estimated to have cost the airlines several billion dollars in additional expense. Delays also drive the need for extra gates and ground personnel and impose costs on airline customers (including shippers) in the form of lost productivity, wages and goodwill.
<br>
<br>

Assuming 47 dollars per hour(1) as the average value of a passenger’s time, flight delays are estimated to have cost air travelers billions of dollars. FAA/Nextor estimated the annual costs of delays (direct cost to airlines and passengers, lost demand, and indirect costs) in 2018 to be 28 billion dollars.
<br>
<br>

(1) FAA-recommended value: https://www.faa.gov/regulations_policies/policy_guidance/benefit_cost/media/econ-value-section-1-tx-time.pdf

### Finding "Direct Aircraft Operating Cost per Block Minute" for Atlanta airport in 2018

In [164]:
df

FL_DATE ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  CANCELLED  DIVERTED  \
3       2018-01-01    ATL  ORD          1748          3          0         0   
4       2018-01-01    ATL  IAD           943         22          0         0   
5       2018-01-01    ATL  EWR          2011         45          0         0   
7       2018-01-01    ATL  EWR          1329          1          0         0   
12      2018-01-01    ATL  IAH          1738          9          0         0   
...            ...    ...  ...           ...        ...        ...       ...   
387563  2018-12-31    ATL  ILM          2119          2          0         0   
387569  2018-12-31    ATL  PHX          1740         76          0         0   
387570  2018-12-31    ATL  CLT          1220         20          0         0   
387571  2018-12-31    ATL  PHX          1125         59          0         0   
387573  2018-12-31    ATL  ORD          1737          3          0         0   

       AIR_TIME  DISTANCE  CARRIER_DELAY  WEATHER_DELAY  LATE_AIRCRAFT_DELAY  \
3            95       606              0              0                    0   
4            70       534              0              0                    0   
5            94       746              0              0                    0   
7           132       746              0              0                    0   
12          106       689              0              0                    0   
...         ...       ...            ...            ...                  ...   
387563       59       377              0              0                    0   
387569      240      1587             75              0                    0   
387570       56       226              0              0                   20   
387571      251      1587              4              0                   55   
387573       92       606              0              0                    0   

        Temperature  Dew Point  Humidity  Wind Speed  Pressure  Precipitation  \
3                27         10        49          13     29.35            0.0   
4                19          6        57          15     29.36            0.0   
5                 0          0         0           0      0.00            0.0   
7                23          8        53          13     29.34            0.0   
12               27         10        49          13     29.35            0.0   
...             ...        ...       ...         ...       ...            ...   
387563           67         61        81          14     28.94            0.0   
387569           68         59        73          13     28.91            0.0   
387570           66         63        90           9     28.94            0.1   
387571           66         63        90           9     28.94            0.1   
387573           70         59        68          17     28.92            0.0   

            Condition                 time  
3       Partly Cloudy  2018-01-01 17:51:00  
4          Light Snow  2018-01-01 10:05:00  
5                   0  2018-01-01 20:56:00  
7              Cloudy  2018-01-01 13:30:00  
12      Partly Cloudy  2018-01-01 17:47:00  
...               ...                  ...  
387563  Mostly Cloudy  2018-12-31 21:21:00  
387569  Mostly Cloudy  2018-12-31 18:56:00  
387570     Light Rain  2018-12-31 12:40:00  
387571     Light Rain  2018-12-31 12:24:00  
387573  Mostly Cloudy  2018-12-31 17:40:00  

[141182 rows x 20 columns]

In [165]:
df["DELAY_COST"]=df.DEP_DELAY*74.18 

In [166]:
df.DELAY_COST.value_counts()

74.18       10303
148.36       7602
222.54       6697
296.72       6179
370.90       5607
            ...  
84416.84        1
51258.38        1
39241.22        1
48291.18        1
62904.64        1
Name: DELAY_COST, Length: 720, dtype: int64

In [167]:
df.DELAY_COST.sum()

335816717.36

In [168]:
df_weather_delay=df[df['WEATHER_DELAY'] >0]
df_weather_delay

FL_DATE ORIGIN DEST  CRS_DEP_TIME  DEP_DELAY  CANCELLED  DIVERTED  \
15      2018-01-01    ATL  DSM           950         68          0         0   
34      2018-01-01    ATL  LFT          2042         39          0         0   
69      2018-01-01    ATL  CSG          1050        102          0         0   
163     2018-01-01    ATL  MGM          1025         10          0         0   
169     2018-01-01    ATL  AGS          2116         28          0         0   
...            ...    ...  ...           ...        ...        ...       ...   
387279  2018-12-31    ATL  OAJ          2220         16          0         0   
387321  2018-12-31    ATL  AGS           855         75          0         0   
387496  2018-12-31    ATL  XNA          1051        104          0         0   
387509  2018-12-31    ATL  ABY          1342         97          0         0   
387539  2018-12-31    ATL  GSP          2230         75          0         0   

       AIR_TIME  DISTANCE  CARRIER_DELAY  WEATHER_DELAY  LATE_AIRCRAFT_DELAY  \
15          113       743              0             68                    0   
34           78       503              0             26                    0   
69           25        83              0            102                    0   
163          30       147              0             10                    0   
169          28       143              0             19                    1   
...         ...       ...            ...            ...                  ...   
387279       63       399              0             16                    0   
387321       27       143              0             75                    0   
387496       95       589              0             90                    0   
387509       32       145              0             71                   25   
387539       42       153              0             75                    0   

        Temperature  Dew Point  Humidity  Wind Speed  Pressure  Precipitation  \
15               19          6        57          15     29.36            0.0   
34               22          8        55          13     29.38            0.0   
69               19          6        57           9     29.39            0.0   
163              19          6        57          15     29.36            0.0   
169              22          8        55          13     29.38            0.0   
...             ...        ...       ...         ...       ...            ...   
387279           68         61        78          14     28.94            0.0   
387321           59         58        96           6     29.03            0.0   
387496           66         63        90           9     28.94            0.1   
387509           70         62        76          14     28.92            0.0   
387539           66         63        88           9     28.94            0.0   

            Condition                 time  DELAY_COST  
15         Light Snow  2018-01-01 10:58:00     5044.24  
34               Fair  2018-01-01 21:21:00     2893.02  
69               Fair  2018-01-01 12:32:00     7566.36  
163        Light Snow  2018-01-01 10:35:00      741.80  
169              Fair  2018-01-01 21:44:00     2077.04  
...               ...                  ...         ...  
387279         Cloudy  2018-12-31 22:36:00     1186.88  
387321            Fog  2018-12-31 10:10:00     5563.50  
387496     Light Rain  2018-12-31 12:35:00     7714.72  
387509  Mostly Cloudy  2018-12-31 15:19:00     7195.46  
387539           Rain  2018-12-31 23:45:00     5563.50  

[8408 rows x 21 columns]

In [169]:
df_weather_delay.DELAY_COST.value_counts()

2373.76     112
1631.96     108
1854.50     106
2299.58     104
2225.40     100
           ... 
64462.42      1
89461.08      1
37683.44      1
30413.80      1
43543.66      1
Name: DELAY_COST, Length: 536, dtype: int64

In [170]:
df_weather_delay.DELAY_COST.sum()

60281635.2

## The total flight delay cost in year 2018 was 335,816,717.36 dollars, out of which 60,281,635.2 dollars were lost due to weather delays